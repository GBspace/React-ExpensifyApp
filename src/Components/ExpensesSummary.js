import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

export class ExpenseSummary extends React.Component{
    render(){
        const expenseLabel = this.props.expenseCount === 1 ? 'expense' : 'expenses';
        return(
            <div>
            
                <h1>Viewing {this.props.expenseCount} {expenseLabel} totalling to {this.props.expenseTotal} </h1>
                
                
            </div>
        );
    };
};



const mapStateToProps = (state)=>{
    let visibleExpenses = selectExpenses(state.expense,state.filters);
    return{
        expenseCount: visibleExpenses.length,
        expenseTotal: expensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);