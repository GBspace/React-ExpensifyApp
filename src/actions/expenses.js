import uuid from 'uuid';
import database from '../firebase/firebase';

//Normal redux steps
//component call action generator
//action generator return object
//component dispatches object
//redux store changes

//Redux with firebase steps
//component call action generator
//action generator returns function
//component dispatches function --> not supported by redux
//fucntion runs (it has the ability to dispatch other actions and do whatever it wants)


export const addExpense = (expense) => ({
type: 'ADD_EXPENSE',
expense
});

export const startAddExpense = (expenseData = {})=>{
    return (dispatch)=>{
        const {
            desc = '',
            note = '' , 
            amount = 0, 
            createAt = 0
        } = expenseData;
        const expense = {desc,note,amount,createAt };
     return database.ref('expenses')
        .push(expense)
        .then((ref)=>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

export const removeExpense = ({id} = {})=>({
type: 'REMOVE_EXPENSE',
id

});

export const editExpense = (id,updates)=>({
type: 'EDIT_EXPENSE',
id,
updates
});
