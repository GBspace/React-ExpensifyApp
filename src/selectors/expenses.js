//get visibile expenses
import moment from 'moment';
import { create } from 'domain';

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense)=>{
        const createAtMoment = moment(expense.createAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment,'day'): true;
        const endDateMatch = endDate? endDate.isSameOrAfter(createAtMoment,'day'): true;
        const textMatch = expense.desc.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
        // return textMatch;
        // return startDateMatch && endDateMatch;
    }).sort((a,b)=>{
        if(sortBy==='date') {
            return a.createAt < b.createAt ? 1 : -1;
        }
        else if(sortBy==='amount'){
            return a.amount > b.amount ? 1 : -1;
        }
    });
};
