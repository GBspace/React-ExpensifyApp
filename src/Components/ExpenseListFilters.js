import React from 'react';
import {connect} from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';


export class ExpenseListFilters extends React.Component{
    state = {
        calanderFocused: null
    };
    onDatesChange = ({startDate,endDate})=>{
        
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calanderFocused)=>{
        this.setState(()=>({
            calanderFocused
        }));
    };

    onTextChange =(e)=>{
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e)=>{
        e.target.value==='date' ? this.props.sortByDate() :
        this.props.sortByAmount();
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
        
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value = 'date' > Date </option>
                    <option value = 'amount'> Amount </option>
                </select>
                <DateRangePicker
                    startDate = {this.props.filters.startDate}
                    startDateId= "start"
                    endDate = {this.props.filters.endDate}
                    endDateId = "end"
                    onDatesChange = {this.onDatesChange}
                    focusedInput={this.state.calanderFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                /> 
            </div>
        );
    };
};



const mapStateToProps = (state) =>{
    return{
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch)=>({
    setTextFilter: (text)=>dispatch(setTextFilter(text)),
    sortByDate: ()=>dispatch(sortByDate()),
    sortByAmount: ()=>dispatch(sortByAmount()),
    setEndDate: (endDate)=>dispatch(setEndDate(endDate)),
    setStartDate: (startDate)=>dispatch(setStartDate(startDate))
});

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);