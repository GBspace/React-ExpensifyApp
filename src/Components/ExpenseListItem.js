import React from 'react';
import {removeExpense} from '../actions/expenses';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';


export const ExpenseListItem = ( {id,desc,amount,createAt}) => {
    return (
        <div>
            <Link to={`/edit/${id}`}> Edit </Link>  
            <h3>{desc} </h3>
            <p> {amount} -- {createAt}</p>
        </div>
    );
};



export default connect()(ExpenseListItem);;