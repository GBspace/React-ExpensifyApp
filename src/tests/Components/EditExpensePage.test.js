import {shallow} from 'enzyme';
import React from 'react';
import {EditExpensePage} from '../../Components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense,startRemoveExpense,history,wrapper,expense;

beforeEach(()=>{
    editExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(<EditExpensePage 
                       editExpense={editExpense} 
                       startRemoveExpense={startRemoveExpense}
                       history={history}
                       expense={expenses[0]} />);
});
test('should render edit expense page',()=>{
    expect(wrapper).toMatchSnapshot();

});

test('should handle edit expense',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,expenses[2]);

});

test('should handle startRemoveExpense',()=>{
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id: expenses[0].id});

});