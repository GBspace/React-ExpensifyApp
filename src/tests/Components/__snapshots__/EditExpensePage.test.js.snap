// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 200,
        "createAt": 0,
        "desc": "Gum",
        "id": "1",
        "note": "note1",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     Remove 
  </button>
</div>
`;
